mdb/exchange_server.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2023-2024 Tom Meltzer. See the top-level COPYRIGHT file for
+# details.
+
+import asyncio
+import logging
+import signal
+import ssl
+from typing import Any, Coroutine
+
+from .async_connection import AsyncConnection
+from .messages import DEBUG_CLIENT, MDB_CLIENT, Message
+from .utils import ssl_cert_path, ssl_key_path
+
+logger = logging.getLogger(__name__)
+
+
+class AsyncExchangeServer:
+    def __init__(self, opts: dict[str, Any]):
+        self._init_tls()
+        self.number_of_ranks = opts["number_of_ranks"]
+        self.hostname = opts["hostname"]
+        self.port = opts["port"]
+        self.backend_name = opts["backend"]
+        self.launch_task = opts["launch_task"]
+        self.debuggers: list[AsyncConnection] = []
+        logger.info(f"echange server started :: {self.hostname}:{self.port}")
+
+    def _init_tls(self) -> None:
+        # fergus: i made no changes here other than paths
+        context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
+        context.load_cert_chain(
+            ssl_cert_path(),
+            ssl_key_path(),
+        )
+        # add these two lines to force check of client credentials
+        context.verify_mode = ssl.CERT_REQUIRED
+        context.load_verify_locations(ssl_cert_path())
+        self.context = context
+
+    async def handle_connection(
+        self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter
+    ) -> None:
+        # fergus: what you probably want to do here is have a seperate
+        # connection class that holds onto the connection reader/writer and
+        # implements your handler methods
+
+        # no try/except clause needed as the asyncio server does that for us
+        conn = AsyncConnection(reader, writer)
+        try:
+            msg = await conn.recv_message()
+        except Exception as e:
+            print(e)
+        logger.info(
+            "exchange server received {%s} from {%s}.",
+            msg.msg_type,
+            msg.data["from"],
+        )
+
+        # checkhov's loop
```

Comment:
```suggestion
        # chekhov's loop
```
I am shocked I misspelled his name